### PR TITLE
lazy decorators

### PR DESCRIPTION
**Pull request by scopatz**
Thanks for the idea @melund!

### CODE REVIEW

**Comment by melund** [comment]
Cool. It looks awesome.

**Comment by laerus** [comment]
is there a way to solve  #1392 without calling `__getattribute__` somehow?

**Comment by scopatz** [comment]
@laerus, I think that issue is unrelated to this one

**Comment by scopatz** [comment]
Ok!  I have finished messing around with this branch.  Now everything in the main xonsh dir should be lazy.  Sorry it is such a huge change.  Most of the line diffs are simple changes to pyghooks.

**Comment by laerus** [comment]
Great work @scopatz ! i've been playing with this branch and everything seems to be working nice.

**Comment by melund** [comment]
Great. Looks good to me as well. Love the decorators.

File: xonsh/proc.py [hunk 1]
```diff
@@ -12,46 +12,49 @@
...
+import functools
+import threading
+import importlib
+import subprocess
+import collections
+import collections.abc as abc
 
 from xonsh.tools import (redirect_stdout, redirect_stderr, ON_WINDOWS, ON_LINUX,
                          fallback, print_exception, XonshCalledProcessError)
+from xonsh.teepty import TeePTY
```
Comment (melund) [inline xonsh/proc.py]: @scopatz . Sorry I didn't notice this in the review. 

This change causes xonsh to fail on windows when `XONSH_DEBUG` is set. TeePTY imports some pacakges which are linux only

Comment (scopatz) [inline xonsh/proc.py]: Do you know which ones?  It seems like the right fix is to make those imports explicitly lazy / non-windows.

**Comment by melund** [comment]
termios is not available. Also termios is also imported by both tty and pty

**Comment by scopatz** [comment]
OK, fix in #1417

**Pull request merged**